src/c/autoevent.c

```diff
@@ -41,6 +43,48 @@ static void edgex_autoimpl_release (void *p)
   free (ai);
 }
 
+//helper function to compare values with threshold
+static bool values_exceed_threshold (const devsdk_commandresult *newvals, const devsdk_commandresult *oldvals, int nvals, double threshold, iot_logger_t *logger)
+{
+  iot_log_debug (logger, "Comparing values against threshold: %f", threshold);
+  if(!newvals || !oldvals || threshold == 0.0)
+  {
+    iot_log_debug (logger, "No threshold set or new/old values are null, publishing event.");
+    return true;
+  }
+  bool publish = false;
+  for (int i = 0; i < nvals; i++)
+  {
+    if(newvals[i].value != NULL && oldvals[i].value != NULL){
+      iot_data_t *curr_val_cast = iot_data_transform(newvals[i].value, IOT_DATA_FLOAT64);
+      iot_data_t *prev_val_cast = iot_data_transform(oldvals[i].value, IOT_DATA_FLOAT64);
+      if(curr_val_cast != NULL || prev_val_cast != NULL)
+      {
+        double curr_val = iot_data_f64(curr_val_cast);
+        double prev_val = iot_data_f64(prev_val_cast);
+        iot_log_debug (logger, "Values of index %d: current=%f, previous=%f", i, curr_val, prev_val);
+        if (fabs(curr_val - prev_val) > threshold)
+        {
+          iot_log_debug (logger, "Value change %f exceeds threshold %f publishing event.", fabs(curr_val - prev_val), threshold);
+          publish = true;
+        }
+      }else{
+        if(!iot_data_equal(newvals[i].value, oldvals[i].value))
+        {
+          iot_log_debug (logger, "Non-numeric value changed, publishing event.");
+          publish = true;
+        }
+      }
+      iot_data_free (curr_val_cast);
+      iot_data_free (prev_val_cast);
+      return publish;
```

Comment:
if publish is false we still need to check the remaining values. So, `if (publish) { return true; }`